### PR TITLE
Integrating brushing effects into the plots.

### PR DESCRIPTION
**Pull request by LTLA**

### CODE REVIEW

File: R/plotting.R [hunk 1]
```diff
@@ -48,14 +48,49 @@ names(.all_labs_values) <- .all_aes_names
...
   # Store the ggplot commands
-  cmds$todo[["ggplot"]] <- sprintf("ggplot(plot.data, %s) +", .build_aes(color = color_set))
-  cmds$todo[["point"]] <- "geom_point(size = 1.5) +"
+  cmds$todo[["ggplot"]] <- "ggplot() +"
+
+  # Implementing the brushing effect.
+  if (brush_set) {
+    brush_effect <- param_choices[[.brushEffect]]
+    if (brush_effect==.brushColorTitle) {
+      cmds$todo[["brush_color"]] <- sprintf(
```
Comment (LTLA) [inline R/plotting.R]: Does the order of `geom_point` calls have any effect? If it's anything like how base `points` behaves, I guess we should probably favour plotting of the brushed points on top of the other points.

Comment (kevinrue) [reply R/plotting.R]: yep, the order matter, the last layer wins, i'll check again now

Comment (kevinrue) [reply R/plotting.R]: argh.. obviously i coded it the wrong way around. fixing now

File: R/codetrack.R [hunk 2]
```diff
@@ -1,28 +1,32 @@
...
+    "## Just copy them in a live session of R where your SingleCellExperiment object is",
+    "## ... and if you want to adjust something, just edit the corresponding lines!",
+    "",
+    "## All commands below refers to your `SingleCellExperiment` object as `se`:",
+    sprintf("se <- %s", se_name),
+    "")
   for (i in seq_len(nrow(aobjs))) {
-    tracked_code <- c(tracked_code, 
+    tracked_code <- c(tracked_code,
                       paste0("## ",paste0(aobjs$Type[i],"_",aobjs$ID[i])),
```
Comment (LTLA) [inline R/codetrack.R]: I would use the `.decode_panel_name` to give them a name that is consistent with the UI.

File: R/codetrack.R [hunk 3]
```diff
@@ -1,28 +1,32 @@
...
 
   # storing to a text character vector
-  tracked_code <- c("## Here's the list of commands to generate the plots you created using iSEE",
-                    "## Just copy them in a live session of R where your SingleCellExperiment object is",
-                    "## ... and if you want to adjust something, just edit the corresponding lines!",
-                    "")
+  tracked_code <- c(
+    "## Here's the list of commands to generate the plots you created using iSEE",
+    "## Just copy them in a live session of R where your SingleCellExperiment object is",
+    "## ... and if you want to adjust something, just edit the corresponding lines!",
```
Comment (LTLA) [inline R/codetrack.R]: Careful with the wording here; the edits will get overwritten upon closing the modal. We should be emphasizing read-only status where possible.

File: R/plotting.R [hunk 4]
```diff
@@ -88,16 +88,12 @@ names(.all_labs_values) <- .all_aes_names
...
-      x = NA_character_, # or param_choices[[.redDimXAxis]], if shown
-      y = NA_character_, # or param_choices[[.redDimYAxis]], if shown
+      x = sprintf("Dimension %s", param_choices[[.redDimXAxis]]),
+      y = sprintf("Dimension %s", param_choices[[.redDimYAxis]]),
       color = color_label
     )
   if (!is.null(labs_add)){
-    cmds$todo[["labs"]] <- .build_labs(
-      x = NA_character_, # or param_choices[[.redDimXAxis]], if shown
-      y = NA_character_, # or param_choices[[.redDimYAxis]], if shown
```
Comment (LTLA) [inline R/plotting.R]: I guess you should add it here as well?

**Comment by LTLA** [comment]
Y should always be numeric for a violin plot, so no need for that check. On the other hand, X should never be numeric, otherwise it would have become a scatter plot instead.

**Comment by kevinrue** [comment]
which one is 'that' check?

**Comment by LTLA** [comment]
The `if (numeric_axes["x"]){` and `if (numeric_axes["y"]){` in `.violin_plot()`.

**Comment by kevinrue** [comment]
final checks? (bis)

**Comment by LTLA** [comment]
I can't check on my current computer, so I'll just ask: is the full range of the x-axis preserved automatically when brushing to restrict? I didn't see a `droplevels=FALSE` for the lost levels; or is this default?

**Comment by kevinrue** [comment]
Nope, it's not. Diving back in there!

**Comment by LTLA** [comment]
It may (not sure) be necessary to `as.factor` any character vectors to group by, in order for a hypothetical `droplevels=FALSE` setting to work.

**Comment by kevinrue** [comment]
Right. I thought that was already the case. I can't find this `drop=FALSE` thing anymore. I think I confused with `?ggplot2::facet_grid`
Unrelated: are we going to merge before we solve the `validate(FALSE, "ARGHH!")` situation, or do we still want to resolve it on this branch?

**Comment by LTLA** [comment]
Yes, ignore the scream, I'll deal with it sometime this week.

For any others listening in on this: I should just emphasize that preservation of the violin plot x-coordinates is important if the violin plot is itself a transmitter. This is because `shiny::brushedPoints` converts groupings into x-coordinates via `as.integer(as.factor(...))`. If the number of levels being shown is not the same as that in `plot.data`, it's possible that the wrong points will be transmitted.

To resolve this, we have two options. The most obvious is to ensure we're always plotting the same number of levels in the violin plot, which is the preferred option. The other option is to subset `plot.data` manually. The second is a bit easier to do, but it does mean that the plot will change quite dramatically when you change the upstream inputs. It will probably also render existing brushes meaningless.

**Comment by kevinrue** [comment]
I knew I didn't dream it: `scale_x_discrete(drop=FALSE)`
One those those `...` hiddenoptions .

**Comment by LTLA** [comment]
A kludgey way to do it would be to spike in points for the lost levels, with y-values set to some very large negative number, and then force the y-axis limits back to the original range. This would make sure that the missing factors appear, while maintaining the plot limits (and no one will be none the wiser...)

**Comment by kevinrue** [comment]
No need after my last update: this is readily supported by ggplot

**Comment by LTLA** [comment]
I wonder whether our repeated `geom_quasirandom` calls in `.violin_plot` are correct. Will it re-calculate the density using only the subsetted points (wrong) or use the density from the violin plots (right)?

**Comment by kevinrue** [comment]
If that is indeed the case, I would only see one solution: overplotting . 
Meaning that one layer would  include all points with the plot’s colouring. To ensure.the density is correct.
The subsetted points would then go in a different later, only for highlighting.
Something to look into tomorrow.

**Comment by LTLA** [comment]
I'm not sure even overplotting would help, as the second layer would recalculate the densities if called via `geom_quasirandom`. Perhaps we should just set `color` and `alpha` directly in `plot.data` and hope that the `geom_*` functions respond to them. We can also reorder the rows so that those with `BrushBy=TRUE` are last (thus guaranteeing that they are plotted on top of the others).

File: R/codetrack.R (hunk 3 above)
Comment (federicomarini) [reply R/codetrack.R]: I'll make it clearer then: the adjustments are to be thought once the user copies, executes, sees something needs updates, and edits the corresponding lines in his/her code *after* using `iSEE`

Comment (federicomarini) [reply R/codetrack.R]: Sorry for picking up on the comment later. I see now the line is gone anyway, and you updated that in a clean way

File: vignettes/iSEE_vignette.Rmd [hunk 5]
```diff
@@ -135,5 +134,15 @@ file.copy(from = temp_file, to = newLocation)
 
 # FAQ
 
-*Coming soon.*
+**Q: Can you implement a 'Copy to clipboard' button in the code editor?**
```
Comment (federicomarini) [inline vignettes/iSEE_vignette.Rmd]: 😄

Comment (kevinrue) [reply vignettes/iSEE_vignette.Rmd]: let's see how many users don't read the vignette ;)

**Pull request merged**